### DISCUSSION TITLE
Slow filltoborder

### DISCUSSION THREAD

**Discussion by dcsobral** [ImageMagick/ImageMagick#2458] [Help]
I'm processing some height maps, monochrome images where the pixel value represents an elevation, to create a mask with water locations, but processing it is incredibly slow compared to anything else I've ever done with Imagemagick.

The command I'm running is this:
```bash
        convert -size "${DIM}" -depth 16 gray:dtm.raw -flip \
                -depth 8 \
                -threshold "$threshold" -write mpr:mask \
                -monitor -draw "@${file}" \
                mpr:mask -compose minus_src -composite \
                -alpha on -transparent black \
                "${MASK_IMG}" -compose src-over -composite \
                "${TMPIMG}"
```
The images are usually 8192x8192 in size. `MASK_IMG` is the result of the previous iteration, and initialized with this:
```bash
convert -size "${DIM}" xc:black -alpha on -transparent black "${MASK_IMG}"
```
But slowness comes from the drawing directives. Here's a sample:
```
push defs
push clip-path "-4096,4096,-4096,4096"
push graphic-context
rectangle 0,0 8191,8191
pop graphic-context
pop clip-path
pop defs
fill white
border-color white
push graphic-context
clip-path url(#-4096,4096,-4096,4096)
color 7301,5218 filltoborder
color 7285,5228 filltoborder
color 7302,5227 filltoborder
color 7295,5229 filltoborder
color 7301,5210 filltoborder
color 7294,5213 filltoborder
pop graphic-context
```
This is the output from timing its run:
```
negate image[mask]: 8191 of 8192, 100% complete
negate image[mask]: 8191 of 8192, 100% complete
negate image[mask]: 8191 of 8192, 100% complete
negate image[mask]: 8191 of 8192, 100% complete
negate image[mask]: 8191 of 8192, 100% complete
negate image[mask]: 8191 of 8192, 100% complete
negate image[mask]: 8191 of 8192, 100% complete
Composite/Image[mask]: 8191 of 8192, 100% complete
set transparent color in image[mask]: 8191 of 8192, 100% complete
Composite/Image[mask]: 8191 of 8192, 100% complete
save image[tmp-prefabs-Nitrogen-8192.png]: 8191 of 8192, 100% complete

real    0m37.229s
user    0m46.672s
sys     0m8.094s
```
Each one of those repeated "negate image" lines takes a long time to be produced. The actual area being draw is rather small, and, in fact, I believe it's fully filled by the first "color" directive. This is the resulting mask:
```
$ convert mask-prefabs-Nitrogen-8192.png -format %c histogram:info:-
  67106863: (  0,  0,  0,  0) #00000000 graya(0,0)
      2001: (255,255,255,255) #FFFFFFFF graya(255,1)
```
I believe either I'm doing something wrong, or there's a much better way to do this, but I can't think of anything. Can anyone suggest a way of improving it?

**Comment by urban-warrior** [comment]
The FillToBorder operation has to search an area and all of its connected neighbors.  As it is a floodfill operator, it can be a time consuming op particularly as your image size increases.  The algorithm is threaded so you will get a speed-up as the # of cores on your system increases.
